domain/learningandworkprogress/src/test/kotlin/uk/gov/justice/digital/hmpps/domain/learningandworkprogress/review/service/ReviewScheduleDateCalculationServiceWithEarliestScheduleDateTest.kt

```diff
@@ -0,0 +1,323 @@
+package uk.gov.justice.digital.hmpps.domain.learningandworkprogress.review.service
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.Arguments
+import org.junit.jupiter.params.provider.MethodSource
+import uk.gov.justice.digital.hmpps.domain.learningandworkprogress.review.ReviewScheduleCalculationRule
+import uk.gov.justice.digital.hmpps.domain.learningandworkprogress.review.ReviewScheduleWindow
+import uk.gov.justice.digital.hmpps.domain.learningandworkprogress.review.SentenceType
+import uk.gov.justice.digital.hmpps.domain.randomValidPrisonNumber
+import java.time.LocalDate
+import java.util.stream.Stream
+
+/**
+ * Unit test class for [ReviewScheduleDateCalculationService] that uses a date as the earliest date from which
+ * all calculated schedule dates are set.
+ *
+ * The concept of the earliest schedule is specifically to support creating Review Schedules before the official go live
+ * on 2025-04-1
+ * After 2025-04-01, once the feature has gone live, the optional constructor arg and associated code will be removed,
+ * at which point this unit test can be removed.
+ */
+class ReviewScheduleDateCalculationServiceWithEarliestScheduleDateTest {
+  private val dateCalculationService = ReviewScheduleDateCalculationService(APRIL_FIRST)
+
+  @ParameterizedTest(name = "[{index}] {0}")
+  @MethodSource("determineReviewScheduleCalculationRule_testCases")
+  fun `should determine review schedule calculation rule`(
+    scenario: String,
+    releaseDate: LocalDate?,
+    sentenceType: SentenceType,
+    isReadmission: Boolean,
+    isTransfer: Boolean,
+    expectedReviewScheduleCalculationRule: ReviewScheduleCalculationRule,
+  ) {
+    // Given
+
+    // When
+    val actual = dateCalculationService.determineReviewScheduleCalculationRule(
+      PRISON_NUMBER,
+      sentenceType,
+      releaseDate,
+      isReadmission,
+      isTransfer,
+    )
+
+    // Then
+    assertThat(actual).isEqualTo(expectedReviewScheduleCalculationRule)
+  }
+
+  @ParameterizedTest(name = "[{index}] {0}")
+  @MethodSource("calculateReviewScheduleWindow_testCases")
+  fun `should calculate a review schedule window`(
+    scenario: String,
+    releaseDate: LocalDate,
+    expectedReviewScheduleWindow: ReviewScheduleWindow?,
+  ) {
+    // Given
+
+    // When
+    val reviewScheduleCalculationRule = dateCalculationService.determineReviewScheduleCalculationRule(
+      PRISON_NUMBER,
+      SentenceType.SENTENCED,
+      releaseDate,
+      false,
+      false,
+    )
+    val actual = dateCalculationService.calculateReviewWindow(reviewScheduleCalculationRule, releaseDate)
+
+    // Then
+    assertThat(actual).isEqualTo(expectedReviewScheduleWindow)
+  }
+
+  // Calculate a review schedule window as if the calculation was being performed on the earliest calculation date (April 1st)
```

Comment:
This test, and the method source, contain specific scenarios as worked out and expressed by the BA